File: backend/app/schemas/dataset.py
from __future__ import annotations

from typing import Any, Dict, List

from pydantic import BaseModel


class DatasetMetadata(BaseModel):
    """
    Metadata about the uploaded dataset.
    """

    filename: str
    extension: str
    rows: int
    columns: int
    column_names: List[str]
    memory_usage_mb: float
    missing_values: int
    duplicate_rows: int
    upload_time: str


class DatasetProfile(BaseModel):
    """
    Dataset profiling information.
    """

    rows: int
    columns: int
    numeric_columns: List[str]
    categorical_columns: List[str]
    datetime_columns: List[str]
    boolean_columns: List[str]
    missing_values: Dict[str, int]
    unique_values: Dict[str, int]
    data_types: Dict[str, str]


class DatasetAnalysisResult(BaseModel):
    """
    Result returned by DatasetService after loading a dataset.
    """

    metadata: DatasetMetadata
    profile: DatasetProfile
    statistics: Dict[str, Any]


class UploadResponse(BaseModel):
    """
    API response for dataset upload.
    """

    success: bool
    message: str
    metadata: DatasetMetadata
    profile: DatasetProfile
    statistics: Dict[str, Any]


__all__ = [
    "DatasetMetadata",
    "DatasetProfile",
    "DatasetAnalysisResult",
    "UploadResponse",
]

File: frontend/src/types/dataset.ts
export interface DatasetMetadata {
  filename: string;
  extension: string;
  rows: number;
  columns: number;
  column_names: string[];
  memory_usage_mb: number;
  missing_values: number;
  duplicate_rows: number;
  upload_time?: string;
  filepath?: string;
}

export interface ColumnDetail {
  name: string;
  dtype: string;
  non_null_count: number;
  null_count: number;
  null_percentage: number;
  unique_count: number;
  sample_values?: (string | number | boolean | null)[];
}

export interface DatasetProfile {
  columns: ColumnDetail[];
  numeric_columns: string[];
  categorical_columns: string[];
  datetime_columns: string[];
  missing_values_by_column: Record<string, number>;
  data_types: Record<string, string>;
  head?: Record<string, any>[];
}

export interface DatasetResponse {
  success: boolean;
  message: string;
  metadata: DatasetMetadata;
  profile?: DatasetProfile;
  statistics?: Record<string, any>;
  preview?: Record<string, any>[];
}

export interface CleaningOptions {
  filename?: string;
  missing_strategy?: "drop" | "fill_mean" | "fill_median" | "fill_mode" | "fill_value";
  fill_value?: string | number;
  remove_duplicates?: boolean;
  columns_to_drop?: string[];
}

export interface ChartConfig {
  chart_type: string;
  x_column: string;
  y_column?: string;
  title?: string;
  theme?: string;
  color?: string;
  width?: number;
  height?: number;
  dpi?: number;
}

export interface MLPipelineConfig {
  filename?: string;
  target_column: string;
  feature_columns?: string[];
  problem_type: "classification" | "regression";
  model_type: string;
  test_size?: number;
}

export interface ReportConfig {
  filename?: string;
  title?: string;
  include_charts?: boolean;
  include_ai_summary?: boolean;
}
